### ISSUE TITLE
JSON::Any#inspect should include JSON::Any in its output

### ISSUE THREAD

**Issue by Djok39** [crystal-lang/crystal#6213]
https://play.crystal-lang.org/#/r/4bfg
```
require "json"

json = JSON.parse %( {"array":[null]} )
puts "no result" unless elem = json["array"].first?
puts "First element: #{elem}"
puts "First element: is nil" if elem==nil
```
Crystal 0.24.1 (2018-02-11)
LLVM: 5.0.1
Default target: x86_64-pc-linux-gnu

**Comment by bcardiff** [comment]
@Djok39 I don't follow the issue here. `elem==nil` is returning true.
`#{elem}` is printing an empty string.

On 0.25.0 you will need to replace the `.first?` to `[0]?`.
And you can use `elem.inspect` if you want to see explicitly a `nil`.

```crystal
require "json"

json = JSON.parse %( {"array":[null]} )
puts "no result" unless elem = json["array"][0]?
puts "First element: #{elem.inspect}"
puts "First element: is nil" if elem==nil
```

**Comment by Djok39** [comment]
program should print "no result", because the result in the condition is nil.
but [0]? works correctly.

**Comment by RX14** [comment]
It's because you don't have `nil`, you have `JSON::Any(nil)`. 

https://play.crystal-lang.org/#/r/4bhb

The problem is that `inspect` doesn't make this explicit, the inspect output should contain `JSON::Any`.

**Comment by asterite** [comment]
Yes, `inspect` and `to_s` should use `JSON::Any`, so if you have:

```crystal
require "json"

value = JSON.parse("[1, 2, 3]")
p value
````

you get this as output:

```
JSON::Any([JSON::Any(1_i64), JSON::Any(2_i64), JSON::Any(3_i64)])
```

See how nicely all those `JSON::Any` occurences fit with those `_i64` occurences.

Just kidding... the solution was already given, `elem` is `JSON::Any`, you should compare its raw value with `nil`.

**Comment by straight-shoota** [comment]
I'm pretty sure we can limit this to the outermost JSON::Any 😄

**Comment by Djok39** [comment]
Or make special if/unless operators for JSON...

btw, i agree, that .inspect should show real state.

**Comment by RX14** [comment]
You could have some fiber-local state in `inspect` to only print the `JSON::Any` at the top level. We already have recursion detection for pretty print.

**Comment by ysbaddaden** [comment]
To be honest, the naive approach looks fine to me. It avoids the illusion that the wrapped value is `[nil, 1]` when each array value is actually wrapped. Even if noisy, being explicit helps to understand the actual data structure (yes, it's ugly).

```crystal
JSON::Any([JSON::Any(Nil), JSON::Any(1)])
```

**Comment by straight-shoota** [comment]
IMO this is unnecessarily noisy. The outer `JSON::Any` wrapper sufficiently expresses that every value inside must be wrapped in `JSON::Any` as well. Repeating that again and again doesn't make it any clearer, just more annoying.

For me, the only reason for being extra expressive would be if it resulted in actual Crystal code that could be copy & pasted to recreate the value exactly. Something like `JSON::Any.new([JSON::Any.new(nil), JSON::Any.new(1)])`.
Not sure if that would be very useful, though. Stringified JSON is already a solid data exchange format, so we could just use that.
